Skyline/work/20250627 python installer virtual environments

dshteyn opened the pull request:
Fix for hanging execution when installing python virtual environments.

brendanx67 commented on pwiz_tools/Skyline/Model/Tools/PythonInstaller.cs:
```diff
@@ -1076,13 +1077,13 @@ public RunGetPipScriptTask(PythonInstaller installer)
             // CONSIDER(brendanx): Path quoting?
-            var cmdLine = TextUtil.SpaceSeparate(PythonInstaller.BasePythonExecutablePath,
-                PythonInstaller.GetPipScriptDownloadPath);
+            var cmdLine = TextUtil.SpaceSeparate(TextUtil.Quote(PythonInstaller.BasePythonExecutablePath),
```
Note that the function prototype:

public static string Quote(this string text)

The "this" means you can just use:

PythonInstaller.BasePythonExecutablePath.Quote()

brendanx67 commented on pwiz_tools/Shared/CommonUtil/SystemUtil/PathEx.cs:
```diff
@@ -172,11 +172,9 @@ public static String GetDirectoryName(String path)
+            // Note: Using this function requires an already
             // fully qualified path.
-            if (!Equals(path, Path.GetFullPath(path)))
+            if (!Path.IsPathRooted(path))
```
My read of the documentation is that this does not guarantee a fully qualified path. A rooted path can be relative, e.g. C:Document - rooted by drive but relative on directory, or \Documents - rooted on path, but not drive. I think you likely need to write a new PathEx.IsFullyQualified() function for this which starts simple and eventually calls Path.GetFullPath() in an exception handler, because it can throw if path contains invalid characters. All pointed out by ChatGPT. You can start by asking it to write this function for you.

dshteyn replied:
Added code and associated tests for PathEx.IsFullyQualifiedPath().

chambm replied:
Wow, today I learned... I've thought Windows rooted paths were always absolute for decades. This differs from POSIX where a path is rooted if it starts with a slash, i.e. at the root, and must therefore be absolute.

brendanx67 commented on pwiz_tools/Skyline/Model/Tools/PythonInstaller.cs:
```diff
@@ -144,7 +144,7 @@ internal List<PythonTaskBase> GetPythonTasks(bool? haveNvidiaGpu = false)
 
             return tasks;
         }
- 
```
Please revert unnecesary whitespace and formatting-only changes. These make it more challenging to find the important part of your commit without turning off whitespace diffs.

dshteyn replied:
Apologies for the whitespace and formatting changes no reverted.

brendanx67 reviewed (changes requested):
The added quoting seems good. I assume there was an issue with Path.GetFullPath() throwing? I think the replacement for this code needs to be more thorough than just checking IsPathRooted().

dshteyn commented:
> The added quoting seems good. I assume there was an issue with Path.GetFullPath() throwing? I think the replacement for this code needs to be more thorough than just checking IsPathRooted().

Switched to using PathEx.IsPathFullyQualified and added unit tests.

nickshulman commented on pwiz_tools/Shared/CommonUtil/SystemUtil/PathEx.cs:
```diff
@@ -172,14 +172,50 @@ public static String GetDirectoryName(String path)
+        /// <summary>
+        /// Returns a true if the path is fully qualified
+        /// </summary>
+        /// <param name="path">Path to convert to long-path syntax. This must be a fully qualified path.</param>
```
This part of the comment isn't correct.
The `<param>` tags are not required in these comments so you can just remove it.

nickshulman commented on pwiz_tools/Skyline/Test/LongPathDirectoryTest.cs:
```diff
@@ -30,6 +30,142 @@ namespace pwiz.SkylineTest
+        private const string PREFIX_LONG_PATH = @"\\?\"; // Assuming this is the long-path prefix
+
+        [TestMethod]
+        public void IsPathFullyQualified_ValidDriveLetterPath_ReturnsTrue()
```
I would recommend combining all of the tests in this class into a single function with multiple Assert statements in it.

nickshulman commented on pwiz_tools/Skyline/Test/LongPathDirectoryTest.cs:
```diff
@@ -30,6 +30,142 @@ namespace pwiz.SkylineTest
+            string path = @"C:\Folder\Subfolder";
+            bool result = PathEx.IsPathFullyQualified(path);
+
+            Assert.IsTrue(result, "Expected C:\\Folder\\Subfolder to be fully qualified.");
```
I would recommend:
`Assert.IsTrue(PathEx.IsPathFullyQualified(path), "{0} should be fully qualified", path);`

nickshulman reviewed (commented):
Could we write a class called something like "ArgumentList" which is just a list of raw strings.
When ArgumentList is asked for the full command line, it looks at each argument and decides whether the argument needs quoting based on its value (does the value contain spaces quotes or other special characters).
That seems a lot simpler than having to decide whether to call "quote" before adding each argument to a list.
